File: python/prob14.py
#!/usr/bin/env python

cache = {}

def total_steps(start):
    if start in cache:
        return cache[start]

    steps = 1
    n = start
    while n != 1:
        steps += 1
        if n % 2 == 0:
            n /= 2
        else:
            n = 3*n + 1
    cache[start] = steps
    return steps
    


def prob14(limit):
    max_n = 1
    max_steps = 1

    for i in xrange(1, limit):
        if total_steps(i) > max_steps:
            max_n = i
            max_steps = total_steps(i)
    return max_n, max_steps

if __name__ == '__main__':
    import sys

    print prob14(int(sys.argv[1]))

File: java/prob14.java
import java.util.HashMap;

public class prob14 {
    static HashMap<Long,Long> cache = new HashMap<Long,Long>();
    static {
        cache.put(1L, 1L);
    }

    static long totalSteps(long start) {
        long i = start;
        long steps = 0;
        while (true) {
            if (cache.containsKey(i)) {
                long allSteps = steps + cache.get(i);
                cache.put(start, allSteps);
                return allSteps;
            } else if (i % 2 == 0) {
                steps++;
                i /= 2;
            } else {
                steps++;
                i = 3*i + 1;
            }
        }
    }

    static long prob14(long limit) {
        long max_n = 1;
        long max_steps = 1;
        for (long i=1; i<limit; i++) {
            long steps = totalSteps(i);
            if (steps > max_steps) {
                max_steps = steps;
                max_n = i;
            }
        }
        return max_n;
    }

    public static void main(String[] args) {
        System.out.println(prob14(Long.parseLong(args[0])));
    }
}
